compiler/rustc_codegen_llvm/src/intrinsic.rs

```diff
@@ -203,6 +203,41 @@ impl<'ll, 'tcx> IntrinsicCallMethods<'tcx> for Builder<'_, 'll, 'tcx> {
             }
             sym::unlikely => self
                 .call_intrinsic("llvm.expect.i1", &[args[0].immediate(), self.const_bool(false)]),
+            sym::select_unpredictable => {
+                let cond = args[0].immediate();
+                assert_eq!(args[1].layout, args[2].layout);
+                let select = |bx: &mut Self, true_val, false_val| {
+                    let result = bx.select(cond, true_val, false_val);
+                    bx.set_unpredictable(&result);
+                    result
+                };
+                match (args[1].val, args[2].val) {
+                    (OperandValue::Ref(true_val), OperandValue::Ref(false_val)) => {
+                        assert!(true_val.llextra.is_none());
+                        assert!(false_val.llextra.is_none());
+                        let llty = if let PassMode::Cast { cast: ty, pad_i32: _ } = &fn_abi.ret.mode
+                        {
+                            ty.llvm_type(self)
+                        } else {
+                            args[1].layout.llvm_type(self)
+                        };
+                        let ptr = select(self, true_val.llval, false_val.llval);
+                        let out = self.load(llty, ptr, true_val.align);
```

Comment:
We should not perform aggregate loads. I think what you want to do here is something like `OperandValue::Ref(ptr).store(self, result)`, which lowers to a memcpy.

Reply:
I've updated it to do that. I'm still not entirely sure if it's doing the right thing, but at least it does the same thing as `black_box` now.